HeterogeneousCore/SonicTriton/interface/TritonMemResource.h

```diff
@@ -0,0 +1,90 @@
+#ifndef HeterogeneousCore_SonicTriton_TritonMemResource
+#define HeterogeneousCore_SonicTriton_TritonMemResource
+
+#include <string>
+#include <memory>
+
+#include "HeterogeneousCore/SonicTriton/interface/grpc_client_gpu.h"
+
+//forward declaration
+template <typename IO>
+class TritonData;
+struct cudaIpcMemHandle_st;
+typedef cudaIpcMemHandle_st cudaIpcMemHandle_t;
+
+//base class for memory operations
+template <typename IO>
+class TritonMemResource {
+public:
+  TritonMemResource(TritonData<IO>* data, const std::string& name, size_t size, bool canThrow);
+  virtual ~TritonMemResource() {}
+  uint8_t* addr() { return addr_; }
+  size_t size() const { return size_; }
+  bool status() const { return status_; }
+  virtual void copy(const void* values, size_t offset) {}
+  virtual void copy(const uint8_t** values) {}
```

Comment:
What is the difference between the two `copy()` functions?

Reply:
The first is used for input, and the second is used for output. (I tried to have a unified interface, but there are weird rules and limitations regarding `void*` and `const [type]**`, so this seemed simpler.) I'll add some brief comments.

Reply:
I'd find something along `copyTo()` and `copyFrom()` more helpful than comments since the other actor in the copy is always the object itself.

On the "copy from", who owns the pointer returned by pointer? Something else than the caller? How about
```cpp
// returns non-owning pointer
const uint8_t* copyFrom() {}
```
?

Reply:
Thinking more, if the "copy from" does not return an owning pointer, why would it be a "copy"?

Reply:
For input, `copy()` takes the content of `values` and copies it into the gRPC buffer or shared memory region. For output, `copy()` takes the content of the gRPC buffer or shared memory region and copies it into `values`. (In some cases, this isn't really a copy but rather just changing where a pointer points, but in other cases, there is actually a copy involved.)

These are only used internally by TritonData, so the naming probably isn't very important. I would propose `copyInput()` and `copyOutput()`, which are a little clearer to me.

Returning a pointer works for `copyOutput()`, so I'll update the interface to do that.

Reply:
Yes, that's correct.